### DISCUSSION TITLE
Why mid-train is removed from repo?

### DISCUSSION THREAD

**Discussion by leolbyy** [karpathy/nanochat#497] [Q&A] [answered]
From the commits history I can see the reason for removing mid-train is becuase using bos-best-fit-dataloader in base_train, But i don't quite understand how this is related...

**Comment by leolbyy** [comment]
I see the current chat_sft is very similari to previous mid_train, while the original chat_sft seems to be nuked.

Can anyone help me understand the motivation behind this?

**Comment by karpathy** [comment, marked as answer]
The commit mentioned - with the introduction of the BOS-aligned data loader there isn't much difference between the two. So I wanted to delete some code. The BOS thing was the major difference.

**Reply by leolbyy** [reply]
Thank you for the reply!

I may be confused about one thing

> there isn't much difference between the two

Does "the two" here means mid_train & chat_sft?

Because it seems like what happened is that
1. chat_sft is nuked
2. mid_train is renamed to chat_sft

**Comment by gpu-poor** [comment]
@leolbyy seems to have a point.

 I think in this change/commit discards one training paradigm which uses mask from tokeniser ( 0 -> user tokens,  1 -> assistant tokens) and use those to compute loss. this was ensured  in the old chat_sft.py that loss will be computed and backproped only for the tokens spoken by assistant.  In the updated/new  chat_sft.py,  this does not seems to be the case, it is basically same as old mid_train.py

**Comment by ggilley** [comment]
I also thought that there were two different purposes served by mid-train and sft. The mid-train was to teach the model the patterns in a bulk-load way. Then the SFT was to refine the model's output to understand expected lengths and reinforce desired outputs.

I think the packing in the new SFT is great. I had raised a concern earlier about the truncation that's going on.

However, everything I've heard says that you want the patterns all seen in pre/mid-train otherwise SFT encourages the model to hallucinate. Now, it could be at this scale for this purpose that doesn't really matter.

So, seems like the experiment to try is to basically turn chat_sft into two passes. The first that does the pattern training, then the second is to do the loss and backprop only for the assistant tokens. And see if that makes a difference in the scores. (and how it "feels" to chat with the model)

**Comment by gpu-poor** [comment]
Made a PR for the same,  

Did an experiment with masked vs non-masked loss on chat_sft.py and turns out masked loss is better. 

https://github.com/karpathy/nanochat/pull/582 -> for more detail 

chat_eval results 

## Baseline 

```
Final: 1307/2376 (55.01%)
ARC-Easy accuracy: 55.01%
Final: 514/1172 (43.86%)
ARC-Challenge accuracy: 43.86%
Final: 5122/14042 (36.48%)
MMLU accuracy: 36.48%
Rank 0 | 104/1319 (7.88%)
==================================================
Final: 104/1319 (7.88%)
GSM8K accuracy: 7.88%
Rank 0 | 5/164 (3.05%)
==================================================
Final: 5/164 (3.05%)
HumanEval accuracy: 3.05%
Rank 0 | 253/256 (98.83%)
==================================================
Final: 253/256 (98.83%)
SpellingBee accuracy: 98.83%
```

## Masked loss
```
Final: 1379/2376 (58.04%)
ARC-Easy accuracy: 58.04%
Final: 553/1172 (47.18%)
ARC-Challenge accuracy: 47.18%
Final: 5100/14042 (36.32%)
MMLU accuracy: 36.32%
Rank 0 | 127/1319 (9.63%)
==================================================
Final: 127/1319 (9.63%)
GSM8K accuracy: 9.63%
Rank 0 | 19/164 (11.59%)
==================================================
Final: 19/164 (11.59%)
HumanEval accuracy: 11.59%
Rank 0 | 256/256 (100.00%)
==================================================
Final: 256/256 (100.00%)
SpellingBee accuracy: 100.00%
```
